filesystem.go

```diff
@@ -0,0 +1,150 @@
+package gluster
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/gluster/gogfapi/gfapi"
+	"gopkg.in/src-d/go-billy.v4"
+)
+
+const (
+	strNoFileOrDir = "no such file or directory"
+
+	defaultDirectoryMode = 0755
+	defaultCreateMode    = 0666
+)
+
+var _ billy.Basic = new(FS)
+
+// FS manages the filesystem of a gluster volume. It implements
+// billy.Basic.
+type FS struct {
+	v *gfapi.Volume
+}
+
+// New creates a new FS connecting it to the specified host and volume.
+func New(host, volume string) (*FS, error) {
+	vol := new(gfapi.Volume)
+	err := vol.Init(volume, host)
+	if err != nil {
+		return nil, err
+	}
+
+	err = vol.Mount()
+	if err != nil {
+		return nil, err
+	}
+
+	g := &FS{v: vol}
+	return g, nil
+}
+
+// Close unmounts the gluster volume associated to the FS.
+func (g *FS) Close() error {
+	return g.v.Unmount()
+}
+
+// Create implements billy.Basic interface.
+func (g *FS) Create(filename string) (billy.File, error) {
+	if err := g.createDir(filename); err != nil {
+		return nil, err
+	}
+
+	f, err := g.v.Create(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewFile(filename, f, os.O_RDWR), nil
+}
+
+// Open implements billy.Basic interface.
+func (g *FS) Open(filename string) (billy.File, error) {
+	return g.OpenFile(filename, os.O_RDONLY, 0)
+}
+
+// OpenFile implements billy.Basic interface.
+func (g *FS) OpenFile(
+	filename string,
+	flag int,
+	perm os.FileMode,
+) (billy.File, error) {
+	if flag&os.O_CREATE != 0 {
```

Comment:
`O_CREAT                          = 0x40`

maybe for safety I would always do:
`flags & O_CREAT  == O_CREAT`